### ISSUE TITLE
RekeyFrag doesn't want to be added to KeyStore after being reconstituted from bytes

### ISSUE THREAD

**Issue by jMyles** [nucypher/nucypher-kms#126]
It seems that after using `RekeyFrag.from_bytes`, it can't be added to the `KeyStore`:

```
        kfrag_data = b''
        if sig:
            kfrag_data += sig
>       kfrag_data += kfrag.id + kfrag.key
E       elliptic_curve.Error: invalid type specified.
```

It's trying here to add kfrag.id + kfrag.key, which are ints, to the kfrag_data, which is bytes of the sig.

**Comment by michwill** [comment]
They probably are elliptic_curve.ec elements (not quite ints, and that's good - otherwise the addition would wrongly work). They indeed have to be serialized to bytes themselves first using ec.serialize

**Comment by jMyles** [comment]
Right - so that's what the `__bytes__` method does.  However, ultimately the problem is: the tests (like `test_kfrag_sqlite`) show `id` and `key` as `bytes`, but `from_bytes` turns them into `ec.element` objects.

The whole thing is difficult to read and needs to be simplified. 

Thus, you can't actually take a `RekeyFrag` that is created as depicted in the tests and use its `__bytes__`, because you get:

```
 File "<ipython-input-2-298e876b5ec9>", line 1, in <module>
    ec.serialize(self.id) + ec.serialize(self.key)
SystemError: <built-in function serialize> returned NULL without setting an error
```

**Comment by michwill** [comment]
Oh wow, that is really strange. Probably I'll poke it a little bit.. which test does it fail / do you have this test committed to your branch?

**Comment by jMyles** [comment]
OK, there's a commit that shows the error.

**Comment by michwill** [comment]
Ah I see. Well, we should always set id and key in `RekeyFrag` to deserialized values (and `RekeyFrag` is merely a namedtuple). I see that in `test_keyfrag_sqlite` these attributes are set to bytes which eventually fails (not unexpectedly).

Where in the actual non-test code do we set `id` and `key` of `RekeyFrag` to bytes?

**Comment by michwill** [comment]
Ah wait, it's not namedtuple...

**Comment by michwill** [comment]
Ok, `kfrag.id` and `kfrag.key` are expected to be always EC elements, not bytes. So, we shouldn't do `kfrag.id + kfrag.key`. Where do we do it?

Instead we should do `kfrag_data += bytes(kfrag)` or something

**Comment by michwill** [comment]
That's actually how it is done in `keystore.py` right now...

**Comment by jMyles** [comment]
Yeah, that's the change I just made.  I figured that was the obvious thing to do.

So, sanity check to affirm that `id` and `key` are ec.elements?  That will break some of @tuxxy's work, which assumes they're bytes.

**Comment by michwill** [comment]
They are, as far as I am aware of. Which @tuxxy work assumes they're bytes?

**Comment by cygnusv** [comment]
Probably not directly related to this, but this shows a problem of current dependency with [Charm-Crypto's code to handle EC points and scalars](https://github.com/nucypher/nucypher-pre-python/blob/master/npre/elliptic_curve). I don't like how both EC points and scalars are represented as "EC Elements", it's prone to confusion. Scalars are represented internally as OpenSSL BigNumbers, while points are, well, OpenSSL EC Points. Perhaps that difference should be clearer, maybe by some naming convention for starters, but also by refactoring current Charm-Crypto code, if posible.

**Comment by jMyles** [comment]
@cygnusv Heck yeah man.  Agree on all points 100%.

@michwill: Here's an example of using bytes for id and key:

https://github.com/nucypher/nucypher-kms/blob/master/tests/keystore/test_keystore.py#L81

You can see that `add_kfrag` only works if they're bytes.

**Comment by michwill** [comment]
Very good point! One obvious thing to use for that refactoring would be `PyOpenSSL`

**Comment by jMyles** [comment]
Now we're talking.  :-)

**Comment by tuxxy** [comment]
Yeah, @cygnusv points would be fixed in that rewrite. I'd plug Stanford's PBC library, but I think the Python3 bindings for it got deprecated due to someone not maintaining them unfortunately.

**Comment by jMyles** [comment]
Fixed in #124.
